### PR TITLE
feat: add rejected_message support for plugin request-validation

### LINKED ISSUE #5108: request help: Curtomized response message in request-validation plugin
### Issue description

I am enabling the request-validation plugin on a route

curl http://127.0.0.1:9080/apisix/admin/routes/35 -H 'X-API-KEY: xxxxxxxxxxxxxxxxxxx' -X PUT -d '
{
    "uri": "/products/create",
    "plugins": {
        "request-validation": {
            "body_schema": {
                "type": "object",
                "required": ["productName", "price"],
                "properties": {
                    "productName": {"type": "string"},
                    "price": {"type": "number"}
                }
            }
        }
    },
    "upstream": {
        "service_name": "PRODUCTSSERVICE",
        "type": "roundrobin",
        "discovery_type": "eureka"
    }
}'

When I test it using following command - 
curl http://127.0.0.1:9080/products/create  -X POST -d '{"product-Name":"Laptop","pri-ce":12345.00}'

I receive the below message - 
property "price" is required

Can we change this default message with a customized json response

### Environment

 apisix version (cmd: apisix version): 2.8
 OS (cmd: uname -a): 20.04.1-Ubuntu
 OpenResty / Nginx version (cmd: nginx -V or openresty -V): openresty/1.19.3.2
 etcd version, if have (cmd: run curl http://127.0.0.1:9090/v1/server_info to get the info from server-info API): 3.4.0
 apisix-dashboard version, if have: NA
 the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
 uarocks version, if the issue is about installation (cmd: luarocks --version): 3.4.0

### PR DESCRIPTION
**Pull request by leslie-tsang**
### What this PR does / why we need it:

Fix #5108 

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by leslie-tsang** [comment]
@tokers Any comment?

File: apisix/plugins/request-validation.lua [hunk 1]
```diff
@@ -17,27 +17,20 @@
...
+local io            = io
 local req_read_body = ngx.req.read_body
 local req_get_body_data = ngx.req.get_body_data
 
 local schema = {
     type = "object",
+    properties = {
+        header_schema = {type = "object"},
+        body_schema = {type = "object"},
+        rejected_message = {type = "string"}
```
Comment (tokers) [inline apisix/plugins/request-validation.lua:29]: How about limiting its maximum length?

Comment (leslie-tsang) [reply apisix/plugins/request-validation.lua:29]: Good idea.

File: docs/en/latest/plugins/request-validation.md [hunk 2]
```diff
@@ -39,10 +39,11 @@ For more information on schema, refer to [JSON schema](https://github.com/api7/j
...
 
-| Name          | Type   | Requirement | Default | Valid | Description                |
-| ------------- | ------ | ----------- | ------- | ----- | -------------------------- |
-| header_schema | object | optional    |         |       | schema for the header data |
-| body_schema   | object | optional    |         |       | schema for the body data   |
+| Name             | Type   | Requirement | Default | Valid | Description                |
+| ---------------- | ------ | ----------- | ------- | ----- | -------------------------- |
+| header_schema    | object | optional    |         |       | schema for the header data |
+| body_schema      | object | optional    |         |       | schema for the body data   |
+| rejected_message | string | optional    |         |       | schema for the customize reject message |
```
Comment (tokers) [inline docs/en/latest/plugins/request-validation.md:46]: It's better to remind users that at least one of `header_schema` and `body_schema` should be specified.

File: docs/en/latest/plugins/request-validation.md [hunk 3]
```diff
@@ -82,7 +84,7 @@ curl --header "Content-Type: application/json" \
   http://127.0.0.1:9080/get
 ```
 
-If the schema is violated the plugin will yield a `400` bad request.
+If the schema is violated the plugin will yield a `400` bad request with reject response.
```
Comment (tokers) [inline docs/en/latest/plugins/request-validation.md:87]:
```suggestion
If the schema is violated the plugin will yield a `400` bad request with the rejected response.
```

File: docs/en/latest/plugins/request-validation.md [hunk 4]
```diff
@@ -252,3 +254,30 @@ curl http://127.0.0.1:9080/apisix/admin/routes/5 -H 'X-API-KEY: edd1c9f034335f13
     }
 }
 ```
+
+**Custom reject message:**
```
Comment (tokers) [inline docs/en/latest/plugins/request-validation.md:258]: What about giving a request/response example for the custom rejected message?

File: t/plugin/request-validation.t [hunk 5]
```diff
@@ -1449,3 +1449,147 @@ passed
...
+--- response_body
+passed
+--- error_code chomp
+200
+--- no_error_log
+[error]
+
+
+
+=== TEST 37: add route (test request validation `header_schema.required` failure with custom reject message)
```
Comment (tokers) [inline t/plugin/request-validation.t:1551]: Better add more test cases to test these routes.

File: docs/en/latest/plugins/request-validation.md (hunk 3 above)
Comment (leslie-tsang) [reply docs/en/latest/plugins/request-validation.md:87]: Roger that

File: docs/en/latest/plugins/request-validation.md (hunk 4 above)
Comment (leslie-tsang) [reply docs/en/latest/plugins/request-validation.md:258]: Good suggestion.

File: docs/en/latest/plugins/request-validation.md (hunk 2 above)
Comment (leslie-tsang) [reply docs/en/latest/plugins/request-validation.md:46]: Roger that.

File: t/plugin/request-validation.t [hunk 6]
```diff
@@ -1449,3 +1448,147 @@ passed
...
+--- response_body
+passed
+--- error_code chomp
+200
+--- no_error_log
+[error]
+
+
+
+=== TEST 36: add route (test request validation `body_schema.required` success with custom reject message)
```
Comment (spacewander) [inline t/plugin/request-validation.t:1502]: TEST 35 & 36 are the same?

File: t/plugin/request-validation.t [hunk 7]
```diff
@@ -1449,3 +1448,147 @@ passed
...
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
+--- response_body
+passed
+--- error_code chomp
```
Comment (spacewander) [inline t/plugin/request-validation.t:1495]: The error_code: 200 is checked by default. We don't need to check it explicitly.

File: t/plugin/request-validation.t (hunk 6 above)
Comment (leslie-tsang) [reply t/plugin/request-validation.t:1502]: No, TEST 35 test for `header_schema`, TEST 36 test for `body_schema`, with same properties

File: t/plugin/request-validation.t (hunk 5 above)
Comment (leslie-tsang) [reply t/plugin/request-validation.t:1506]: Done.

File: docs/en/latest/plugins/request-validation.md (hunk 4 above)
Comment (leslie-tsang) [reply docs/en/latest/plugins/request-validation.md:260]: How about deal with it in another PR?

File: t/plugin/request-validation.t (hunk 7 above)
Comment (leslie-tsang) [reply t/plugin/request-validation.t]: Removed.

Comment (leslie-tsang) [reply t/plugin/request-validation.t]: Removed.

File: docs/zh/latest/plugins/request-validation.md [hunk 8]
```diff
@@ -38,10 +38,13 @@ title: request-validation
 
 ## 属性
 
-| Name          | Type   | Requirement | Default | Valid | Description                       |
-| ------------- | ------ | ----------- | ------- | ----- | --------------------------------- |
-| header_schema | object | 可选        |         |       | `header` 数据的 `schema` 数据结构 |
-| body_schema   | object | 可选        |         |       | `body` 数据的 `schema` 数据结构   |
+> 注意, `header_schema` 与 `body_schema` 必须至少填写其中一个
```
Comment (tzssangglass) [inline docs/zh/latest/plugins/request-validation.md:41]:
```suggestion
> 注意, `header_schema` 与 `body_schema` 至少填写其中一个
```

File: t/plugin/request-validation.t (hunk 6 above)
Comment (tzssangglass) [reply t/plugin/request-validation.t:1460]: TEST 35 & 36 which actually checks the schema? Can we learn from this？https://github.com/apache/apisix/blob/6162ca0ef92445fdd15b7a213655887b66cd4eea/t/plugin/request-validation.t#L27-L45

Comment (tzssangglass) [reply t/plugin/request-validation.t:1460]: Can the following check schema be implemented in a similar way?

**Comment by tzssangglass** [comment]
should add test case about hits the route with the request-validation plugin and get the rejected_message configured, such as:https://github.com/apache/apisix/blob/6162ca0ef92445fdd15b7a213655887b66cd4eea/t/plugin/uri-blocker.t#L442-L487

**Comment by tzssangglass** [comment]
BTW, I prefer to change `rejected_message` to `rejected_msg`, to keep similar as other plugins, such as `uri-blocker`, `limit-*`.

File: t/plugin/request-validation.t (hunk 6 above)
Comment (leslie-tsang) [reply t/plugin/request-validation.t:1460]: Sure

**Comment by leslie-tsang** [comment]
> should add test case about hits the route with the request-validation plugin and get the rejected_message configured, such as:
> 
> https://github.com/apache/apisix/blob/6162ca0ef92445fdd15b7a213655887b66cd4eea/t/plugin/uri-blocker.t#L442-L487

Original test case don't cover the hit request as well. Test case issue will be fix in next commit.
Thanks for suggestion.

**Comment by leslie-tsang** [comment]
> should add test case about hits the route with the request-validation plugin and get the rejected_message configured, such as:
> 
> https://github.com/apache/apisix/blob/6162ca0ef92445fdd15b7a213655887b66cd4eea/t/plugin/uri-blocker.t#L442-L487

Done

**Comment by tokers** [comment]
Changes now look good to me but what about adding some test cases to cover the use?

**Comment by leslie-tsang** [comment]
> Changes now look good to me but what about adding some test cases to cover the use?

How about do it in next PR, this PR focus on add feature, next focus on use case.

**Review by spacewander — COMMENTED** [review]
The test should be submitted in the same PR of the feature.

**Comment by leslie-tsang** [comment]
> The test should be submitted in the same PR of the feature.

Test of feature has been added.

File: apisix/plugins/request-validation.lua [hunk 9]
```diff
@@ -94,11 +87,11 @@ function _M.rewrite(conf)
         if not body then
             local filename = ngx.req.get_body_file()
             if not filename then
-                return 500
+                return 500, conf.rejected_msg or nil
```
Comment (spacewander) [inline apisix/plugins/request-validation.lua:90]:
```suggestion
                return 500, conf.rejected_msg
```

The default value of conf.rejected_msg is nil

File: docs/en/latest/plugins/request-validation.md [hunk 10]
```diff
@@ -39,10 +39,13 @@ For more information on schema, refer to [JSON schema](https://github.com/api7/j
...
-| ------------- | ------ | ----------- | ------- | ----- | -------------------------- |
-| header_schema | object | optional    |         |       | schema for the header data |
-| body_schema   | object | optional    |         |       | schema for the body data   |
+> Note that at least one of `header_schema` and `body_schema` must be filled in.
+
+| Name             | Type   | Requirement | Default | Valid | Description                |
+| ---------------- | ------ | ----------- | ------- | ----- | -------------------------- |
+| header_schema    | object | optional    |         |       | schema for the header data |
+| body_schema      | object | optional    |         |       | schema for the body data   |
+| rejected_message | string | optional    |         |       | schema for the customize reject message |
```
Comment (spacewander) [inline docs/en/latest/plugins/request-validation.md:48]:
```suggestion
| rejected_message | string | optional    |         |       | the custom rejected message |
```

File: docs/en/latest/plugins/request-validation.md [hunk 11]
```diff
@@ -252,3 +256,30 @@ curl http://127.0.0.1:9080/apisix/admin/routes/5 -H 'X-API-KEY: edd1c9f034335f13
     }
 }
 ```
+
+**Custom reject message:**
```
Comment (spacewander) [inline docs/en/latest/plugins/request-validation.md:260]:
```suggestion
**Custom rejected message:**
```

File: docs/zh/latest/plugins/request-validation.md [hunk 12]
```diff
@@ -38,10 +38,13 @@ title: request-validation
...
-| ------------- | ------ | ----------- | ------- | ----- | --------------------------------- |
-| header_schema | object | 可选        |         |       | `header` 数据的 `schema` 数据结构 |
-| body_schema   | object | 可选        |         |       | `body` 数据的 `schema` 数据结构   |
+> 注意, `header_schema` 与 `body_schema` 至少填写其中一个
+
+| Name             | Type   | Requirement | Default | Valid | Description                       |
+| ---------------- | ------ | ----------- | ------- | ----- | --------------------------------- |
+| header_schema    | object | 可选        |         |       | `header` 数据的 `schema` 数据结构 |
+| body_schema      | object | 可选        |         |       | `body` 数据的 `schema` 数据结构   |
+| rejected_message | string | 可选        |         |       | 自定义拒绝信息的 `schema` 数据结构  |
```
Comment (spacewander) [inline docs/zh/latest/plugins/request-validation.md:47]:
```suggestion
| rejected_message | string | 可选        |         |       | 自定义的拒绝信息  |
```

File: t/plugin/request-validation.t [hunk 13]
```diff
@@ -1445,7 +1406,255 @@ qr/table expected, got string/
...
+test: abc
+--- error_code: 400
+--- response_body chomp
+customize reject message for header_schema.required
+--- error_log eval
+qr/schema validation failed/
+
+
+
+=== TEST 38: pass TEST 35 rule
```
Comment (spacewander) [inline t/plugin/request-validation.t:1484]: We don't need a pass test for this feature as we don't change it?

File: t/plugin/request-validation.t [hunk 14]
```diff
@@ -1445,7 +1406,255 @@ qr/table expected, got string/
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 36: hit TEST 35 rule with empty header
```
Comment (spacewander) [inline t/plugin/request-validation.t:1460]: Better not use the `TEST xxx` in the title as it might be changed.

Comment (leslie-tsang) [reply t/plugin/request-validation.t:1460]: Make sense. I will fix it.

File: t/plugin/request-validation.t (hunk 13 above)
Comment (leslie-tsang) [reply t/plugin/request-validation.t:1484]: We need to test it to make sure it work as expected

File: t/plugin/request-validation.t (hunk 14 above)
Comment (leslie-tsang) [reply t/plugin/request-validation.t]: @spacewander fixed.

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**